### DISCUSSION TITLE
How to repeat a Learn/Replay macro from repeat_macro()

### DISCUSSION THREAD

**Discussion by grege2** [eteran/nedit-ng#202] [General]
Hi, great idea to start up Github "Discussions".

Here's a question then - I'd like to be able to make a macro, in autoload.nm, to repeat the last "Learn/Replay" keyboard macro N times. This is just to streamline things .. maybe I'm doing a repetitive change to a series of lines, like turning:
```
1 E7D2D1    // hex RRGGBB color
2 D2A9A9
3 BF8687
(hundreds more lines ..)
```
into ..
```
   str += "1: E7D2D1\n";
   str += "2: D2A9A9\n";
   str += "3: BF8687\n";
   ...
```
and I've fumbled my way through a keyboard macro to change one line, with numerous mistakes and corrections, that's ok, efficiency isn't any consideration, and I would like to call repeat_macro("Learn/Replay", 20) from a keystroke, to rapidly work down the list, and near the end I can use Cmd-K to finish the last few lines.

I can't manage to successfully call repeat_macro("Learn/Replay", N), or discover the format that nedit-ng wants by looking in `DialogMacro.cpp` and `macro.cpp`. I'm still working on it, but maybe y'all know the answer. Also, it occurs to me this may not be catered for in Nedit, it's a bit recursive I guess, calling a macro from a macro, kind of.

Thanks !

**Comment by anjohnson** [comment]
I usually do that kind of thing using a Regular Expression in the Replace dialog, I find it more powerful than using macros. For your specific example this works for me:

![image](https://user-images.githubusercontent.com/5453863/102286883-d28cf300-3efe-11eb-805f-93fa58940c43.png)

Start by getting the Find regex working and use parentheses to capture all the parts that you need in your replacement string, then add the replacement side. You can move the focus to the text window and use Undo (Ctrl+Z) to undo a replacement in order to fix and retry it. Once you have a single substitution working properly the "Replace & Find" button allows for manual confirmation that it located an appropriate spot first, or the "Replace all in" Window and Selection buttons let you repeat it much faster.

**Comment by grege2** [comment]
That is indeed more professional, thanks. And it would work better when there are variations in the format of the lines. The step-by-step keyboard macro relies on a completely regular line format. In fact even different trailing blanks often blow it up :=(

**Comment by grege2** [comment]
But, I'm still interested in how to call repeat_macro("Learn/Replay", n), it could be useful in quite a few situations.

**Comment by eteran** [comment]
Well, I'm honestly unsure exactly how that could be used correctly! Truth be told, I'm not personally a heavy macro user. I love nedit for its regex capabilities.

But there may be a way to do what you intend. Hopefully one of us will be able to figure it out :-P.

**Comment by eteran** [comment]
I can say that you can use navigation macros like `process_down()` to move the cursor and `insert_string()` and you can always do the pre-cooked operations in a loop. (which is literally what the repeat_macro stuff does internally anyway.

**Comment by grege2** [comment]
Hmm, I will try to come up with a good use case, where  ' repeat_macro("Learn/Replay", nn) ' would be really convenient.

**Comment by eteran** [comment]
I've been thinking about this one...

It **sounds** like what you want is for whatever is in the "learn/replay" macro buffer, to be able to called by name in a macro itself. Perhaps it could be available as like `replay()` or something. Then you could write:

```
for(i = 0; i < n; i++) {
    replay();
}
```

This would also allow:

```
repeat_macro(n, "replay()");
```

Or something like that, which would enable you to record a macro, and then replay whatever that macro was via another macro `n` times?

(It is admittedly hard to imagine what I would use this for given that there is a method to do this via the dialog if I understand what you're trying to do here).

**Comment by grege2** [comment]
Yes that's pretty much what I was thinking. I'll definitely come up with a good use case soon.

**Reply by grege2** [reply]
Ok, here's something close ..

I have a file encoding an image ..
```
// data/mandrill.jpg

  px = "524A33FF120B00FF3C351BFF4F482CFF766F52FF40391CFF837D5DFF484222FF45411EFF"; acc(px);  // line 0001 of 3122
  px = "4A4E13FF565C1EFF71773BFF666B31FF737842FF6D7441FF4A5020FFA2A87CFF4F5527FF"; acc(px);  // line 0002 of 3122
  px = "818668FF474B30FF2E3219FF454932FF3E422BFF252912FF32361FFF444831FF797F63FF"; acc(px);  // line 0003 of 3122
  px = "98A78AFF55654AFF7E8D76FF45543FFF2E3D2AFF314232FF3D4E3EFF708073FF4E6052FF"; acc(px);  // line 0004 of 3122
  px = "596551FF5F6855FF434C39FF35402FFF4E5948FF404C3EFF3B4739FF606E61FF758A7BFF"; acc(px);  // line 0005 of 3122
```

For some reason I want to reformat it as ..
```
  px2 = "524A33FF 120B00FF 3C351BFF 4F482CFF 766F52FF 40391CFF 837D5DFF 484222FF 45411EFF"; acc2(px);  // line 0001 of 3122
  px2 = "4A4E13FF 565C1EFF 71773BFF 666B31FF 737842FF 6D7441FF 4A5020FF A2A87CFF 4F5527FF"; acc2(px);  // line 0002 of 3122
  etc.
```
Just to give context, these are RGBA image pixels and I'm separating the pixels a bit. It _might_ be doable with a regex. but the simple and intuitive way is just to peck out a keyboard macro for one line, backtrack to correct mistakes, change your mind about what looks good, etc.

Now I want to change all the lines. There are 3122 as shown and it's an irksome process to repeat "Meta K" 3122 times. And, more often I don't know how many lines there are. I'd like to have a key-binding (my own macro binding) for say "replay_keyboard_macro(100)" to get down most of the file in big chunks, and then manually finish the last few lines. I know you can pop the Repeat Macro dialog and enter 3122 (if you know that) or 3000 (if you want to guess) but that seems irksome too, and accidentally going beyond the end of the area you want to modify can scramble following code.

For CG work which I'm doing this kind of repetitive re-formatting comes up a lot. I hope there isn't a method in existing nedit macros that lets me do it how I'd like to ! in which case I will have to eat these words. Anyway that is close to the canonical situation I find myself wanting to have "replay_keyboard_macro(nn)" as a macro function, inside another macro. 

Cheers.

**Reply by eteran** [reply]
LOL, Yea, I'd probably do it with a regex, something like this:

```
replace_all("([A-F0-9]{8})(?!\")", "\\1 ", "regex")
```

**EDIT**: Improved the regex a little bit to do it right on one-shot :-)

**Reply by eteran** [reply]
That being said, you can record a macro, and then select all the lines and choose "repeat in selection". That should do all 3122 lines :-)

**Comment by grege2** [comment]
"Repeat In Selection" is a good thought. Damn. Well I'll keep pondering here.
